homeassistant/components/watts/coordinator.py

```diff
@@ -0,0 +1,137 @@
+"""Data coordinator for Watts Vision integration."""
+
+from __future__ import annotations
+
+from datetime import datetime, timedelta
+import logging
+
+from visionpluspython.client import WattsVisionClient
+from visionpluspython.models import Device
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, FAST_POLLING_INTERVAL, UPDATE_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class WattsVisionHubCoordinator(DataUpdateCoordinator[dict[str, Device]]):
+    """Hub coordinator for bulk device discovery and updates."""
+
+    def __init__(
+        self, hass: HomeAssistant, client: WattsVisionClient, config_entry: ConfigEntry
+    ) -> None:
+        """Initialize the hub coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=UPDATE_INTERVAL),
+            config_entry=config_entry,
+        )
+        self.client = client
+        self._devices: dict[str, Device] = {}
+        self._is_initialized = False
+
+    async def async_config_entry_first_refresh(self) -> None:
+        """Perform initial discovery of devices."""
+        try:
+            await self._discover_devices()
+            self.async_set_updated_data(self._devices)
+        except (ConnectionError, TimeoutError, ValueError) as err:
+            _LOGGER.error("Initial device discovery failed: %s", err)
+            raise UpdateFailed(f"Initial discovery failed: {err}") from err
+
+    async def _discover_devices(self) -> None:
+        """Discover devices from API."""
+        devices_list = await self.client.discover_devices()
+        self._devices = {device.device_id: device for device in devices_list}
+        self._is_initialized = True
+        _LOGGER.info("Initial discovery completed with %d devices", len(self._devices))
```

Comment:
Curious: why do we store the devices both in self._devices and in self.data (self.data is populated by the return value of `_async_update_data` and `self.async_set_updated_data`). I think we can just roll with `self.data` here